cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -835,6 +833,10 @@ Status ParquetFileFragment::SetMetadata(
   DCHECK_EQ(manifest_->descr, original_metadata_->schema())
       << "SchemaDescriptor should be owned by the original FileMetaData";
 
+  if (!physical_schema_) {
```

Comment:
Will this need locking like in the other comment?

Reply:
If you look at the existing uses of `SetMetadata`, they either already took the lock or are calling the method on a brand new fragment instance that cannot possibly be used from another thread yet.